cpp/src/parquet/encoding.cc

```diff
@@ -934,13 +956,34 @@ class DictByteArrayDecoder : public DictDecoderImpl<ByteArrayType>,
     return result;
   }
 
+  int DecodeArrow(int num_values, int null_count, const uint8_t* valid_bits,
+                  int64_t valid_bits_offset,
+                  ::arrow::StringDictionaryBuilder* out) override {
+    int result = 0;
+    PARQUET_THROW_NOT_OK(
+        DecodeArrow(num_values, null_count, valid_bits, valid_bits_offset, out, &result));
+    return result;
+  }
+
   int DecodeArrowNonNull(int num_values,
                          ::arrow::internal::ChunkedBinaryBuilder* out) override {
     int result = 0;
     PARQUET_THROW_NOT_OK(DecodeArrowNonNull(num_values, out, &result));
     return result;
   }
 
+  int DecodeArrowNonNull(int num_values, ::arrow::BinaryDictionaryBuilder* out) override {
+    int result = 0;
+    PARQUET_THROW_NOT_OK(DecodeArrowNonNull(num_values, out, &result));
+    return result;
+  }
+
+  int DecodeArrowNonNull(int num_values, ::arrow::StringDictionaryBuilder* out) override {
+    int result = 0;
+    PARQUET_THROW_NOT_OK(DecodeArrowNonNull(num_values, out, &result));
+    return result;
+  }
```

Comment:
The feeling of code duplication is indeed suspicious. A templated impl class could help

Reply:
I reworked the `ByteArrayDecoder` interface to:

* Use a public interface templated over the builder types
* Introduced a `WrappedBuilderInterface` as a sort of shim layer between the real builder instance and the one used within the decoder
* Added private virtual interface for decoding that works on a `WrappedBuilderInterface`

Not sure about the names but let me know if you have any more thoughts on this implementation.